extensions/extended-isupport.md

```diff
@@ -0,0 +1,71 @@
+---
+title: "`extended-isupport` Extension"
+layout: spec
+copyrights:
+  -
+    name: "Simon Ser"
+    period: "2024"
+    email: "contact@emersion.fr"
+---
+
+## Notes for implementing work-in-progress version
+
+This is a work-in-progress specification.
+
+Software implementing this work-in-progress specification MUST NOT use the
+unprefixed `extended-isupport` capability name. Instead, implementations SHOULD
+use the `draft/extended-isupport` capability name to be interoperable with
+other software implementing a compatible work-in-progress version. The final
+version of the specification will use unprefixed capability names.
+
+## Introduction
+
+`RPL_ISUPPORT` is used to advertise metadata about the server as key-value
+entries. However, `RPL_ISUPPORT` is only sent by servers after connection
+registration. This undermines the usefulness of `RPL_ISUPPORT`: some of the
+metadata would be useful to clients prior to connection registration. This
+extension adds a mechanism to send `RPL_ISUPPORT` messages before connection
+registration, as well as an end of `RPL_ISUPPORT` list indication.
+
+## Implementation
+
+### `draft/extended-isupport` capability
+
+When the `draft/extended-isupport` capability is enabled by the client, the
+server MUST send one or more `RPL_ISUPPORT` messages (grouped inside a
+`draft/isupport` batch if the `batch` capability is enabled). While
+`draft/extended-isupport` is enabled, the server MAY send updates to the
+key-value entries via subsequent `RPL_ISUPPORT` messages.
+
+The server MAY skip the `RPL_ISUPPORT` replies usually sent when connection
+registration completes, if it already sent all information.
+
+Before connection registration, the server MAY send only a subset of the full
+`RPL_ISUPPORT` list. In that case, the server MUST send a `RPL_ISUPPORT` list
+when connection registration completes with entries previously omitted.
+
+### `draft/isupport` batch
+
+The server MUST group all `RPL_ISUPPORT` messages inside a `draft/isupport`
+batch when the [`batch`][] and `draft/extended-isupport` capabilities are
+enabled.
+
+The batch MUST only contain one or more `RPL_ISUPPORT` messages.
+
+## Examples
+
+Enabling the capability:
+
+    C: CAP LS 302
+    S: :irc.example.org CAP * LS :multi-prefix sasl draft/extended-isupport
+    C: CAP REQ draft/extended-isupport
+    S: :irc.example.org CAP * ACK draft/extended-isupport
+    S: :irc.example.org BATCH +asdf draft/isupport
```

Comment:
This appears to show `BATCH` being sent without the client having enabled the `batch` CAP.

My reading of the current draft is that `BATCH` should only be sent if the client enables the `batch` CAP. Is that the correct understanding? (Another possibility is that requesting `draft/extended-isupport` implies the ability to receive batches, and therefore the server should always send `BATCH`.)

Reply:
Following up on discussion in #ircv3: this spec is similar to [draft/multiline](https://ircv3.net/specs/extensions/multiline) in that it depends on the `batch` CAP. draft/multiline says:

>This specification depends on the [batch](https://ircv3.net/specs/extensions/batch) capability which MUST be negotiated to use multiline messages. The order of capability negotiation is not significant and MUST not be enforced.

I read "MUST not be enforced" in this case to mean that if the client requests `draft/multiline` before `batch`, the server MUST send an `ACK` for `draft/multiline` and record that it was enabled. If the client continues with only `draft/multiline` enabled, server behavior is undefined.

However, the request order is necessarily significant for `draft/extended-isupport` because of the "edge-triggered" design. If the client successfully requests `draft/extended-isupport`, it should receive an 005 burst in response. If the client doesn't have `batch` enabled at that moment, what should the server do? (My vote would be for the server to send the 005 burst without batching.)

Reply:
I don't see why the batch cap couldn't be optional for clients here tbh. Not every client needs the request in a batch.

Reply:
That works for me too --- just want the spec to be explicit about the behavior, and the example to be consistent with that behavior.

Reply:
The examples seems wrong here indeed.

My preference would be to allow `draft/extended-isupport` without `batch` (in which case messages are sent unbatched) and make the order insignificant.

Reply:
If I understand correctly, this is the expected behavior:

```
>>> CAP REQ draft/extended-isupport
:oragono.test CAP * ACK draft/extended-isupport
:oragono.test 005 * AWAYLEN=500 BOT=B CASEMAPPING=ascii CHANLIMIT=#:100 CHANMODES=Ibe,k,fl,CEMRUimnstu CHANNELLEN=64 CHANTYPES=# CHATHISTORY=100 ELIST=U EXCEPTS EXTBAN=,m EXTJWT=1 FORWARD=f :are supported by this server
:oragono.test 005 * INVEX KICKLEN=1000 MAXLIST=beI:60 MAXTARGETS=4 MODES MONITOR=100 MSGREFTYPES=msgid,timestamp NETWORK=OragonoTest NICKLEN=32 PREFIX=(qaohv)~&@%+ RPCHAN=E RPUSER=E STATUSMSG=~&@%+ :are supported by this server
:oragono.test 005 * TARGMAX=NAMES:1,LIST:1,KICK:,WHOIS:1,USERHOST:10,PRIVMSG:4,TAGMSG:4,NOTICE:4,MONITOR:100 TOPICLEN=1000 UTF8ONLY WHOX draft/CHATHISTORY=100 :are supported by this server
>>> CAP REQ batch
:oragono.test CAP * ACK batch
```

This sounds correct to me (at the moment `draft/extended-isupport` is enabled, `batch` is not yet enabled, so the burst is sent unbatched). But I think "the order in which the capabilities are enabled is not significant" is a misleading description of the behavior: the order is significant in that it affects whether the burst is batched.

Reply:
This isn't typically what cap ordering means, afaik it means that

    CAP REQ :batch draft/extended-isupport

is equivalent to

    CAP REQ :draft/extended-isupport batch

Reply:
Hmm, I always understood `REQ` as an atomic and simultaneous request for all the capabilities on the line. Is order ever significant on the same line?

Reply:
Ref https://github.com/ircv3/ircv3-specifications/pull/549#issuecomment-2267452934

Reply:
In multiline the intention is just that a server shouldn't NAK a client that REQs multiline and then batch with two separate REQs.

Reply:
The example has been fixed, and it's been spelled out that the order doesn't matter.